sfwreporting.c

```diff
@@ -0,0 +1,1184 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ *****************************************************************************/
+
+#include "sfwreporting.h"
+
+#include "sfwservice.h"
+#include "sfwsensor.h"
+#include "sfwlogging.h"
+#include "sfwdbus.h"
+#include "utility.h"
+
+/* ========================================================================= *
+ * Constants
+ * ========================================================================= */
+
+#define ENABLE_DEFAULT false
+#define ENABLE_INVALID (-1)
+
+#define DEFAULT_DATARATE (0)
+#define INVALID_DATARATE (-1)
+
+#define DEFAULT_OVERRIDE false
+#define INVALID_OVERRIDE (-1)
+
+/* ========================================================================= *
+ * Types
+ * ========================================================================= */
+
+typedef enum SfwReportingState
+{
+    SFWREPORTINGSTATE_INITIAL,
+    SFWREPORTINGSTATE_DISABLED,
+    SFWREPORTINGSTATE_RETHINK,
+    SFWREPORTINGSTATE_STARTING,
+    SFWREPORTINGSTATE_CONFIGURE,
+    SFWREPORTINGSTATE_STARTED,
+    SFWREPORTINGSTATE_STOPPING,
+    SFWREPORTINGSTATE_STOPPED,
+    SFWREPORTINGSTATE_FAILED,
+    SFWREPORTINGSTATE_FINAL,
+    SFWREPORTINGSTATE_COUNT
+} SfwReportingState;
+
+typedef struct SfwReportingPrivate
+{
+    SfwSensor         *rpt_sensor;
+    gulong             rpt_sensor_changed_id;
+
+    /* State */
+    SfwReportingState  rpt_state;
+    bool               rpt_valid;
+    bool               rpt_active;
+    guint              rpt_eval_state_id;
+
+    /* Fail - Retry */
+    guint              rpt_retry_delay_id;
+
+    /* Start / Stop */
+    bool               rpt_enable_wanted;
+    int                rpt_enable_requested;
+    int                rpt_enable_effective;
+    GCancellable      *rpt_enable_cancellable;
+
+    /* Datarate */
+    double             rpt_datarate_wanted;
+    double             rpt_datarate_requested;
+    double             rpt_datarate_effective;
+    GCancellable      *rpt_datarate_cancellable;
+
+    /* Stand-by override */
+    bool               rpt_override_wanted;
+    int                rpt_override_requested;
+    int                rpt_override_effective;
+    GCancellable      *rpt_override_cancellable;
+
+} SfwReportingPrivate;
+
+struct SfwReporting
+{
+    GObject object;
+};
```

Comment:
Split into public and private parts only make sense if one part is public and the other one is private. If the whole thing is private (defined in the .c file and not accessible from elsewhere) then why not to define a single private structure.

Reply:
IIRC there was some guidance along the line: using public/private data separation can make it easier to make objects inheritable later on or stte.

Reply:
If there are plans to expose the object and especially make it inheritable - then yes, it does make sense.